### PR TITLE
ovirt-imageio-container: tool creation and testing

### PR DESCRIPTION
**Pull request by aesteve-rh**
Create a new tool to start and ease the management of
the containerized imageio.

Currently, it requires to manually create a static ticket,
handle paths, both local and in the container, and
set environment variables, all of that only to start
the imageio server.

We could have a tool that does all of that for you, on
the fly:
```
ovirt-imageio-container --port 8080 disk.img
```
It could also serve using nbd protocol without having to start
`qemu-nbd` manually, and select another ticket.
```
ovirt-imageio-container --port 8080 --backend file disk.img
```

This will also simplify testing container in the CIs.

Depends on: https://github.com/oVirt/ovirt-imageio/pull/145
Signed-off-by: Albert Esteve <aesteve@redhat.com>

### CODE REVIEW

File: examples/conf.d/daemon.conf [hunk 1]
```diff
@@ -2,6 +2,9 @@
 #
```
Comment (nirs) [inline examples/conf.d/daemon.conf:2]: Can you split this change to another pr, so we can merge it now?

**Comment by nirs** [comment]
> To run the container:
> 
>    podman run --publish 8000:8000 ovirt-imageio

This cannot work since we don't have a ticket or an image in the container.
Please update the example for a working server. This should also be documented,
maybe add docs/container.md?

> More work is needed:
> - Inject a ticket and PKI files into the container

PKI can be done later, but we want to have a ticket now.

> - Mount the image to transfer or nbd socket in the container

This is just a mater of documentation - the ticket includes the path of
the image in the container, so only the podman command to mount the image
in the container is missing.

Later we want also to show how to mount nbd socket in the container.

> - Modify the server to load a ticket from file during startup

This is already done

The commit message needs upadte?

File: examples/conf.d/daemon.conf (hunk 1 above)
Comment (aesteve-rh) [reply examples/conf.d/daemon.conf:2]: Sure! Makes sense.

File: containers/centos-8.containerfile [hunk 2]
```diff
@@ -15,6 +15,7 @@ RUN echo v1 \
         libguestfs-tools-c \
         make \
         openssl \
+        podman \
```
Comment (nirs) [inline containers/centos-8.containerfile:18]: Running podman in socket in the CI works?

File: ci/test.sh [hunk 3]
```diff
@@ -29,5 +27,12 @@ fi
 
 env="$1"
 
-make storage
-tox -e flake8,test-$env,bench-$env
+if [ $env = "container" ]; then
+    make container
+    tox -e $env
```
Comment (nirs) [inline ci/test.sh:32]: Can we do this inside tox instead?

For example:

    commands_pre = 
        container: make container

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -15,12 +15,20 @@ jobs:
             env: py36
           - distro: centos-8
             env: py38
+          - distro: centos-8
+            env: container
```
Comment (nirs) [inline .github/workflows/ci.yml:19]: So we double the number of jobs, one job will run the tests, and a second job will run the container tests?

Which version of the container is tested? If we test the container (e.g. alpine) why do we need to run the same tests 4 times with all distros? If the code under test is what we have in the container, we need to test it only once.

Maybe we need another job, running on ubuntu (without docker), building the imageio container (using docker or podman on ubuntu), starting the container using github yaml, and running the relevant tests (on ubuntu), accessing the running container?

But we have to think also about local testing - we can actually start with having only local testing - how a developer will run the tests locally?

I think we need to think more about the big picture and agree on how this should work:
- local testing
- ci testing
- real usage with podman/docker
- real usage in k8s

Comment (aesteve-rh) [reply .github/workflows/ci.yml:19]:
```
Maybe we need another job, running on ubuntu (without docker), building the imageio container (using docker or podman on ubuntu), starting the container using github yaml, and running the relevant tests (on ubuntu), accessing the running container?
```
I considered that too, for the same reason you expose. So far, I am mostly testing what is needed to make it work (like checking dependencies, seeing if the podman approach works in the CIs, etc.). After that, I will refine this and probably follow the approach you mentioned.

Regarding the local test, it should be already possible to run local tests. I have run them. But I need to document it. And sure, let's have a discussion about it.

File: tox.ini [hunk 5]
```diff
@@ -55,6 +62,7 @@ addopts = -rsxX --basetemp=/var/tmp/imageio-tests --strict
 # https://docs.pytest.org/en/latest/mark.html
 markers =
     benchmark
+    container
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply tox.ini:65]: Test under this mark require building the container image beforehand (`make container`), and do not care about code coverage.

[the remark this answers is not in the document]
Comment (nirs) [reply tox.ini:65]: We already have `make images` creating the base image uses in some tests (e.g. backup). Maybe we can have `make container` that will run when running `make`, so no special setup is needed.

So developer will run:

    make
    make check (or tox, tox -e ...)

File: container/Containerfile [hunk 6]
```diff
@@ -10,11 +10,12 @@ RUN /venv/bin/pip install ovirt-imageio.tar.gz
 # Final stage
 
 FROM alpine
+ARG TICKET_PATH=ticket.json
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply container/Containerfile:13]: You are right, `ARG` is not needed as is not used at build time.

The example command should be `--volume test:ticket`, and the default `ARG` is not considered during runtime anyway. It was there just so that the build command didn't complain. I tested and can just remove it.

File: test/containerized_test.py [hunk 7]
```diff
@@ -16,13 +25,113 @@
...
+
+
+@pytest.fixture
+def tmp_ticket(tmpdir, tmp_image):
+    return TestTicket(tmpdir, tmp_image)
+
+
+@pytest.fixture
+def srv(tmpdir, tmp_ticket, tmp_image):
+    cmd = ["podman", "run", "--rm", "--privileged", "-it"]
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply test/containerized_test.py:98]: Apparently it does not work. I was not sure if it would work, but I had hopes. I just had never tried before. That sucks, but I'm not sure what else can be done to test it in the pipelines.

File: ci/test.sh (hunk 3 above)
Comment (aesteve-rh) [reply ci/test.sh]: We could if we add make to the allowlist_externals:
```
[testenv]
allowlist_externals = make
```

File: test/containerized_test.py [hunk 8]
```diff
@@ -16,13 +25,113 @@
...
+    def __init__(self, file_name, folder):
+        self.file_name = file_name
+        self.folder = folder
+
+    @property
+    def path(self):
+        return os.path.join(self.folder, self.file_name)
+
+
+class TestImage:
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply test/containerized_test.py]: That's right, I forgot. In the end, I have removed the classes anyway.

File: test/containerized_test.py [hunk 9]
```diff
@@ -16,13 +25,113 @@
...
+
+SERVER_PORT = 8080
+CONTAINER_IMAGE = "localhost/ovirt-imageio:latest"
+
+
+class ContainerPath:
+
+    def __init__(self, file_name, folder):
+        self.file_name = file_name
+        self.folder = folder
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply test/containerized_test.py]: Wanted to avoid having too many repetitions of the `os.path.dirname()`. And actually looking for a more readable option. But in the current version I opted for global variables.

File: test/containerized_test.py [hunk 10]
```diff
@@ -16,13 +25,113 @@
...
+    cmd.extend(("-v", f"{os.path.dirname(tmp_image.path)}"
+                      f":{tmp_image.container.folder}:Z"))
+    # Ticket path environment.
+    cmd.extend(("--env", f"TICKET_PATH={tmp_ticket.container.path}"))
+    cmd.append(CONTAINER_IMAGE)
+    # Run command.
+    srv_proc = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    # Wait for server to start.
+    time.sleep(2)
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply test/containerized_test.py]: Well, is valid, just not robust enough :D

But I agree, I saw the implementation in `sockutil` and did something similar, but using a connection and probing. Also tried using a `sockutil.TCPAddress` socket, but gave false positives with the containerized server.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (aesteve-rh) [reply .github/workflows/ci.yml]: OK, hit my head against a wall for a while with the ci testing…

Couldn't find a way to make it work so far. The test in this branch works locally, but not in the pipelines. So maybe we can start, as you mentioned, with only local testing. And after that I'll keep trying to make the ci testing run. We would need to discuss how do we want to achieve it.

File: ci/test.sh (hunk 3 above)
Comment (aesteve-rh) [reply ci/test.sh]: I kept it in the `test.sh` script to be consistent with other tests with similar requirements (e.g., `make image` or `make storage`), which are not triggered from inside tox either.

File: test/containerized_test.py [hunk 11]
```diff
@@ -6,41 +6,126 @@
...
+    return Ticket(ticket["uuid"], str(ticket_path))
+
+
+def _wait_for_server(host, port, timeout):
+    start = time.monotonic()
+    deadline = start + timeout
+    conn = http_client.HTTPConnection(host, port)
+    while True:
+        try:
+            conn.connect()
```
Comment (nirs) [inline test/containerized_test.py:76]: This basically does tcp connect without sending any http request, so I wonder why our existing helper for waiting for a socket does not work.

File: test/containerized_test.py [hunk 12]
```diff
@@ -6,41 +6,126 @@
...
+        else:
+            log.debug("Waited for %.6f seconds", time.monotonic() - start)
+            return True
+
+
+@pytest.fixture
+def srv(tmp_ticket, tmp_image):
+    host = "localhost"
+    random_port = testutil.random_tcp_port()
+    cmd = [PODMAN_CMD, "run", "--rm", "--privileged", "-it"]
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply test/containerized_test.py:91]: The server fails to serve the PUT request if it is not privileged:
```
2022-10-07 08:31:53,074 INFO    (MainThread) [server] Starting (hostname=73350289b0ed pid=1, version=2.4.7)                                                                                                                                                                                                                   
2022-10-07 08:31:53,091 INFO    (MainThread) [services] remote.service listening on ('::', 80)                                                                                                                                                                                                                                
2022-10-07 08:31:53,092 INFO    (MainThread) [services] control.service listening on '/run/ovirt-imageio/sock'                                                                                                                                                                                                                
2022-10-07 08:31:53,093 INFO    (MainThread) [server] Initial ticket: /ticket/file.json                                                                                                                                                                                                                                       
2022-10-07 08:31:53,093 ERROR   (MainThread) [server] Server failed                                                                                                                                                                                                                                                           
Traceback (most recent call last):                                                                                                                                                                                                                                                                                            
  File "/venv/lib/python3.10/site-packages/ovirt_imageio/_internal/server.py", line 46, in main                                                                                                                                                                                                                               
    server = Server(cfg, ticket=args.ticket)
  File "/venv/lib/python3.10/site-packages/ovirt_imageio/_internal/server.py", line 140, in __init__
    self._add_ticket(ticket)
  File "/venv/lib/python3.10/site-packages/ovirt_imageio/_internal/server.py", line 195, in _add_ticket
    self.auth.add(self._read_ticket(ticket))
  File "/venv/lib/python3.10/site-packages/ovirt_imageio/_internal/server.py", line 199, in _read_ticket
    with open(ticket, 'r') as f:
PermissionError: [Errno 13] Permission denied: '/ticket/file.json'
```
I tried to give all permissions with `chmod`, but not sure why it still would complain.

[Here](https://github.com/oVirt/ovirt-imageio/blob/master/ovirt_imageio/_internal/backends/__init__.py#L94) it tries to open the ticket using the 'r+' mode, based on the ticket ops… to open the ticket. But the ticket could be always opened as read-only, and maybe that would make it easier to work with them.

File: container/Containerfile [hunk 13]
```diff
@@ -17,4 +17,4 @@ RUN mkdir /run/ovirt-imageio /etc/ovirt-imageio
 COPY --from=build /venv /venv
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply container/Containerfile:17]: `Is this the container/test directory or the main test directory?` I tested on the container folder, and hence the commit message. I'll change the location of the ticket to `test` and fix the commit message.

`Do we need to define environment variable?` Not really, we need to get the variable to the container, either way works. Actually in the test I use the `--env TICKET_PATH=` option.

`ovirt-imageio-container --port 8080 --backend file disk.img` I like it! I think we can aim for this next.

File: test/containerized_test.py [hunk 14]
```diff
@@ -6,41 +6,126 @@
...
+    srv_proc = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    # Wait for server to start.
+    if not _wait_for_server(host, random_port, timeout=10):
+        log.error("Dumping server logs:")
+        log.warning("%s", srv_proc.stdout.read().decode("utf-8"))
+        log.error("%s", srv_proc.stderr.read().decode("utf-8"))
+        pytest.fail("Server could not start")
+    yield Server(host, random_port)
+    srv_proc.terminate()
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply test/containerized_test.py:113]: OK. I'll split the PR and leave this PR for the `ovirt-imageio-container` + tests.

File: test/containerized_test.py (hunk 11 above)
Comment (aesteve-rh) [reply test/containerized_test.py:76]: In my tests the existing helper gave false positives, passing the probing loop and then failing when actually sending the requests. Didn't look too much into it, as this solution seemed appropriate.

File: container/test/ticket.json [hunk 15]
```diff
@@ -0,0 +1,8 @@
+{
+    "uuid": "ticket-id",
+    "size": 6442450944,
+    "url": "file:///images/disk.img",
+    "timeout": 3000,
+    "sparse": true,
+    "ops": ["read", "write"]
+}
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply container/test/ticket.json:8]: Actually, since we already have tickets in the examples folder, they do not differ, and there is no other process (test or tool) that uses it, I think I will just remove it.

File: test/containerized_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nirs) [reply test/containerized_test.py:91]: This is likely caused by forgetting the :Z in

    --volume examples:/ticket:Z

Running privileged container likey hides this issue.

I don't think [opening the backend](https://github.com/oVirt/ovirt-imageio/blob/master/ovirt_imageio/_internal/backends/__init__.py#L94) is related.

**Comment by nirs** [comment]
Note for using file backend:

    ovirt-imageio-container --port 8080 --backend file disk.img

I'm not sure we actually need to support the file backend - better
to use nbd in all cases and start qemu-nbd automatically (inside the container?)
in all cases. This will make it easier to support since qemu-nbd is doing lot of
hard work for us.

**Comment by aesteve-rh** [comment]
Running the nbd server inside the container means we need the path where the image is mounted and the path to the socket, so that it matches the info inside the ticket. Something like this:
```
$ qemu-nbd -t $IMAGE_PATH -k $SOCKET_PATH
```
But I'm not sure whether this can be done like that, since we can only have one `CMD` in the containerfile, this would be a `RUN` command, and we would need the environment variables during build time.

The other option is to pre-define those paths, which is a less flexible solution.

Probably we won't run nbd inside the container easily.

**Comment by nirs** [comment]
> Running the nbd server inside the container means we need the path where the image is mounted and the path to the socket, so that it matches the info inside the ticket. Something like this:
> 
> ```
> $ qemu-nbd -t $IMAGE_PATH -k $SOCKET_PATH
> ```

The actual command is more complicated, see _internal/qemu_nbd.py

> But I'm not sure whether this can be done like that, since we can only have one `CMD` in the containerfile, this would be a `RUN` command, and we would need the environment variables during build time.
> 
> The other option is to pre-define those paths, which is a less flexible solution.
> 
> Probably we won't run nbd inside the container easily.

We need a simple script in the container that will do:
- run qemu-nbd in the background with the path of the mounted image
- create a ticket for the mounted image
- run ovirt-imageio with the ticket path
- wait until ovirt-imageio is ready to serve the image (e.g. wait on the http server)

With this the user can run the container with the image like this:

  podman run -it --rm -e IMAGE=/images/disk.qcow2 -v /path/to/images:/images:Z ovirt-imageio

Or maybe better - if we can pass arguments to CMD:

  podman run -it --rm -v /path/to/images:/images:Z ovirt-imageio /images/disk.qcow2

Stopping the container will kill both qemu-nbd and ovirt-imageio and remove temporary
files and directories created during the transfer.

Writing the script in python using our qemu_nhb and qemu_img modules will be the easiest
way. The script can be packaged as an ovirt-imageioctl command:

    .ovirt-imageioctl serve /path/to/image

This way it can be used also without a container. But we can start with a simple script that will be shipped inside the container, and make it available outside of the container later.

**Comment by nirs** [comment]
@aesteve-rh I think everything we discussed here is already built into the ovirt-imageio
container, so we can close this.

**Comment by aesteve-rh** [comment]
@nirs yes, except the test for the container.

But there is not much left that can be reused anyway. And I'm not sure that we want to go this way anyway.
So let's close the PR.
